### PR TITLE
feat: add H3Regionizer

### LINKED ISSUE #50: Add H3Regionizer
Divides given shape into H3 cells of selected resolution.

### PR DESCRIPTION
**Pull request by Calychas**
Closes #50

### CODE REVIEW

**Comment by Calychas** [comment]
Still needs some tests and an improvement in buffering because it is quite slow

File: srai/regionizers/h3_regionizer.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    into H3 cells - hexagons with pentagons as a very rare exception
+
+    """
+
+    def __init__(self, resolution: int, buffer: bool = True) -> None:
+        """
+        Init H3Regionizer.
+
+        Args:
+            resolution (int): Size of the cells.
```
Comment (RaczeQ) [inline srai/regionizers/h3_regionizer.py:39]: I think we can stick with the word `Resolution` instead of size here, even if it doubles. Size is a complete inversion in this context (higher resolution == smaller size).
```suggestion
            resolution (int): Resolution of the cells. See [1] for a full comparison.
```

And then attach this https://h3geo.org/docs/core-library/restable/ as a reference.

File: srai/regionizers/h3_regionizer.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                H3 Cells (visible on the borders).
+                Turn off for large geometries, as it's computationally expensive.
+                Defaults to True.
+
+        Raises:
+            ValueError: If resolution is not within 0 <= resolution <= 15.
+
+        """
+        if not (0 <= resolution <= 15):
+            raise ValueError(f"Resolution {resolution} is not within 0 <= resolution <= 15.")
```
Comment (RaczeQ) [inline srai/regionizers/h3_regionizer.py:50]: Not sure if it's the logically correct sentence.
```suggestion
            raise ValueError(f"Resolution {resolution} is not within 0 - 15 range.")
```

File: srai/regionizers/h3_regionizer.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        Raises:
+            ValueError: If provided GeoDataFrame has no crs defined.
+
+        """
+        gdf_wgs84 = gdf.to_crs(epsg=4326)
+        # transform multipolygons to multiple polygons
+        gdf_exploded = gdf_wgs84.explode(index_parts=True).reset_index(drop=True)
+
+        h3_indexes = (
+            seq(gdf_exploded["geometry"])
```
Comment (RaczeQ) [inline srai/regionizers/h3_regionizer.py:77]: Nice usage of `pyfunctional`!

File: srai/regionizers/h3_regionizer.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            ValueError: If resolution is not within 0 <= resolution <= 15.
+
+        """
+        if not (0 <= resolution <= 15):
+            raise ValueError(f"Resolution {resolution} is not within 0 <= resolution <= 15.")
+
+        self.resolution = resolution
+        self.buffer = buffer
+
+    def transform(self, gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
```
Comment (RaczeQ) [inline srai/regionizers/h3_regionizer.py:55]: Logically, all required functions should be placed above this function. If it wasn't inside a class, it wouldn't compile because of it.
I'd put `_buffer`, `_gdf_from_h3_indexes`, `_h3_index_to_shapely_polygon` and the rest in the correct order above the `transform` function.

File: srai/regionizers/h3_regionizer.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            h3.Polygon: Converted polygon.
+
+        """
+        exterior = [coord[::-1] for coord in list(polygon.exterior.coords)]
+        interiors = [
+            [coord[::-1] for coord in list(interior.coords)] for interior in polygon.interiors
+        ]
+        return h3.Polygon(exterior, *interiors)
+
+    def _buffer(self, gdf: gpd.GeoDataFrame, gdf_h3: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
```
Comment (RaczeQ) [inline srai/regionizers/h3_regionizer.py:153]: Wouldn't it be faster to buffer the original geometries with `shapely` by an amount equal to half the size of a hex on a given resolution (not sure if half or full radius)?
@simonusher was doing this in his original code in a concise way.

File: tests/regionizers/test_h3_regionizer.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+"""Tests for H3Regionizer."""
+
+import osmnx as ox
+import pytest
+
+from srai.regionizers.h3_regionizer import H3Regionizer
```
Comment (RaczeQ) [inline tests/regionizers/test_h3_regionizer.py:6]: With added import to `srai/regionizers/__init__.py` this can be shortened.
```suggestion
from srai.regionizers import H3Regionizer
```

File: srai/regionizers/voronoi_regionizer.py [hunk 7]
```diff
@@ -76,10 +76,10 @@ def transform(self, gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
             GeoDataFrame with the regionized data cropped using input GeoDataFrame.
 
         Raises:
-            ValueError: If provided geodataframe has no crs defined.
+            ValueError: If provided GeoDataFrame has no crs defined.
 
         """
-        gds_wgs84 = gdf.set_crs(epsg=4326)
+        gds_wgs84 = gdf.to_crs(epsg=4326)
```
Comment (RaczeQ) [inline srai/regionizers/voronoi_regionizer.py:82]: Lol, how did I miss that XD

File: srai/regionizers/h3_regionizer.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        h3_disk = seq(gdf_h3.index).map(h3.grid_disk).flatten().distinct().to_list()
+        gdf_h3_disk = self._gdf_from_h3_indexes(h3_disk)
+
+        gdf_h3_border = (
+            gdf_h3_disk.sjoin(gdf[["geometry"]], predicate="overlaps")
+            .drop(columns="index_right")
+            .drop_duplicates()
+        )
+
+        return pd.concat([gdf_h3, gdf_h3_border], axis=0).drop_duplicates()
```
Comment (RaczeQ) [inline srai/regionizers/h3_regionizer.py:191]: Will `pd.concat` preserve resulting type as a `GeoDataFrame`?

File: examples/regionizers/h3_regionizer.ipynb [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import geopandas as gpd\n",
+    "import osmnx as ox\n",
+    "import matplotlib.pyplot as plt\n",
+    "import h3\n",
+    "from shapely import geometry\n",
+    "from srai.regionizers.h3_regionizer import H3Regionizer"
```
Comment (RaczeQ) [inline examples/regionizers/h3_regionizer.ipynb:22]:
```suggestion
    "from srai.regionizers import H3Regionizer"
```

File: tests/regionizers/test_h3_regionizer.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+@pytest.mark.parametrize(  # type: ignore
+    "geoname,resolution,buffer",
+    [
+        ("Wroclaw, Poland", 7, True),
+        ("Tokyo, Japan", 7, False),
+        ("London, England", 7, True),
+    ],
+)
+def test_transform(geoname: str, resolution: int, buffer: bool) -> None:
```
Comment (RaczeQ) [inline tests/regionizers/test_h3_regionizer.py:17]: Maybe some more tests? Check for example what happens when resolution out of range is passed, and what happens if empty GDF is passed (like in the voronoi regionizer tests).

File: tests/regionizers/test_h3_regionizer.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    Args:
+        geoname (str): name of the place
+        resolution (int): resolution for H3 Regionizer
+        buffer (bool): whether to buffer the H3 cells to fill the borders
+
+    """
+    gdf = ox.geocode_to_gdf(geoname)
+    gdf_h3 = H3Regionizer(resolution, buffer).transform(gdf)
+
+    assert len(gdf_h3) > 0
```
Comment (RaczeQ) [inline tests/regionizers/test_h3_regionizer.py:30]: Maybe another test for asserting that buffered transform covers whole geometry: 
- merge using `_merge_disjointed_gdf_geometries`
- make the difference between the original GDF and buffered hexagonal regions
- make sure that it's empty (`is_empty`)

**Review by RaczeQ — CHANGES_REQUESTED** [review]

File: srai/regionizers/h3_regionizer.py (hunk 4 above)
Comment (Calychas) [reply srai/regionizers/h3_regionizer.py:55]: I think I have another opinion on that. I want the class to be read clearly from top to bottom. Starting with what's the most important (the public API - transform()) and then the implementation details in order of invocation. I understand the logical aspect, that something must be first declared to be used later, but it is not the way we read. I would follow the [stepdown rule](https://marekhudyma.com/code-style/2021/03/02/step-down-rule.html) by the well-known author that shares the surname with the GOT writer xD

File: srai/regionizers/h3_regionizer.py (hunk 5 above)
Comment (Calychas) [reply srai/regionizers/h3_regionizer.py:153]: Yes, it would be a lot faster. I have this implementation myself too, but it is unfortunately not ideal and does not always give correct results. The problem is that the hexagons' edge lengths on the same resolution differ significantly, and the average may be well off compared to the factual edge lengths in obtained hexagons. 
But as I said, I need to work on that a bit more and maybe I will fall back to this solution.

File: srai/regionizers/h3_regionizer.py (hunk 8 above)
Comment (Calychas) [reply srai/regionizers/h3_regionizer.py:191]: I checked and it is in fact GeoDataFrame, but typing of pd.concat() is ofc wrong

File: tests/regionizers/test_h3_regionizer.py (hunk 10 above)
Comment (Calychas) [reply tests/regionizers/test_h3_regionizer.py:17]: Yep, need to add some more tests

File: tests/regionizers/test_h3_regionizer.py (hunk 11 above)
Comment (Calychas) [reply tests/regionizers/test_h3_regionizer.py:30]: Yep, need to add some more tests

File: srai/regionizers/h3_regionizer.py (hunk 5 above)
Comment (piotrgramacki) [reply srai/regionizers/h3_regionizer.py:156]: This may be a silly question, but would it work and be relatively fast to just buffer the original geometry with a big margin (2x hex radius?) and fill it with h3 cells, then just make a spatial join with an original geometry? What am I missing here?

Comment (Calychas) [reply srai/regionizers/h3_regionizer.py:156]: Yes, I think the big margin would actually solve that. I will try to implement that then

File: srai/regionizers/h3_regionizer.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+            .map(self._polygon_shapely_to_h3)
+            .flat_map(lambda polygon: h3.polygon_to_cells(polygon, self.resolution))
+            .to_list()
+        )
+
+        gdf_h3 = self._gdf_from_h3_indexes(h3_indexes)
+        if self.buffer:
+            return self._buffer(gdf_exploded, gdf_h3)
+
+        return gdf_h3.to_crs(gdf.crs)
```
Comment (simonusher) [inline srai/regionizers/h3_regionizer.py:90]: Is there a reason for this to be so complicated? IIRC, my solution to this was just:
1. buffer - 2 * edge length or something,
2. polyfill,
3. sjoin with original polygon to remove non-overlapping hexes

File: srai/regionizers/h3_regionizer.py (hunk 5 above)
Comment (simonusher) [reply srai/regionizers/h3_regionizer.py:156]: Yes, that was my solution to this and I thought we could use that here

File: srai/regionizers/h3_regionizer.py (hunk 4 above)
Comment (RaczeQ) [reply srai/regionizers/h3_regionizer.py:58]: Alright, I can accept that.

**Review by piotrgramacki — APPROVED** [review]
Looks good to me. Nicely done 😊

**Review by piotrgramacki — APPROVED** [review]

**Review by RaczeQ — APPROVED** [review]
Amazing tests! Overall great job.
